crates/turborepo-lib/src/daemon/connector.rs

```diff
@@ -0,0 +1,278 @@
+use std::{
+    path::{Path, PathBuf},
+    process::Stdio,
+    sync::Arc,
+    time::Duration,
+};
+
+use command_group::AsyncCommandGroup;
+use log::{debug, error};
+use notify::{Config, Event, EventKind, RecommendedWatcher, Watcher};
+use sysinfo::{ProcessExt, ProcessRefreshKind, RefreshKind, SystemExt};
+use tokio::{net::UnixStream, sync::mpsc, time::timeout};
+use tonic::transport::Endpoint;
+
+use super::{client::proto::turbod_client::TurbodClient, DaemonClient, DaemonError};
+
+#[derive(Debug)]
+pub struct DaemonConnector {
+    pub dont_start: bool,
+    pub dont_kill: bool,
+    pub pid_file: PathBuf,
+    pub sock_file: PathBuf,
+}
+
+impl DaemonConnector {
+    const CONNECT_RETRY_MAX: usize = 3;
+    const SHUTDOWN_TIMEOUT: Duration = Duration::from_secs(1);
+    const SOCKET_TIMEOUT: Duration = Duration::from_secs(1);
+
+    /// Attempt, with retries, to:
+    /// 1. find (or start) the daemon process
+    /// 2. locate its unix socket
+    /// 3. connect to the socket
+    /// 4. send the 'hello' message, negotiating versions
+    ///
+    /// A new server will be spawned (and the old one killed) if
+    /// dont_kill is unset and one of these cases is hit:
+    /// 1. the versions do not match
+    /// 2. the server is not running
+    /// 3. the server is unresponsive
+    pub async fn connect(self) -> Result<DaemonClient<DaemonConnector>, DaemonError> {
+        for _ in 0..Self::CONNECT_RETRY_MAX {
+            let pid = self.get_or_start_daemon().await?;
+            debug!("got daemon with pid: {}", pid);
+
+            let path = match self.wait_for_socket().await {
+                Ok(p) => p,
+                Err(_) => continue,
+            };
+
+            let conn = Self::get_connection(path.into()).await?;
+            let mut client = DaemonClient {
+                client: conn,
+                connect_settings: (),
+            };
+
+            match (client.handshake().await, &self.dont_kill) {
+                (Ok(_), _) => return Ok(client.with_connect_settings(self)),
+                // should be able to opt out of kill
+                (Err(DaemonError::VersionMismatch), true) => {
+                    return Err(DaemonError::VersionMismatch)
+                }
+                (Err(DaemonError::VersionMismatch), false) => {
+                    self.kill_live_server(client, pid).await?;
+                    continue;
+                }
```

Comment:
Could we hide the matching on `dont_kill` by putting this logic into a `maybe_kill_live_server`?
```suggestion
                Err(DaemonError::VersionMismatch) => {
                    self.maybe_kill_live_server(client, pid).await?;
                    continue;
                }
```